docs: improve write-only attribute example (#36605)
hashicorp/terraform: website/docs/language/resources/ephemeral/index.mdx

> <CodeBlockConfig highlight="11">

Reviewer: @bschaatsbergen First of all thank you so much for this example! I _super_ appreciate it! 

How would you feel about moving the expanded `aws_secretsmanager_secret_version` logic to the bottom of the article in a separate `## Examples` section? I'd like this introduction section pretty basic while explaining how write-only arguments work, and then further down this expanded example on real-life usage would be _amazing_.

```hcl

> ephemeral "random_password" "db_password" {

Author: Added, thanks

  length = 16
}

resource "aws_secretsmanager_secret" "db_password" {
  name = "db_password"
}

resource "aws_secretsmanager_secret_version" "db_password" {
  secret_id                = aws_secretsmanager_secret.db_password.id
  secret_string_wo         = ephemeral.random_password.db_password.result
  secret_string_wo_version = 1
}

ephemeral "aws_secretsmanager_secret_version" "db_password" {

>   secret_id = aws_secretsmanager_secret.db_password.id

Reviewer 2: ```suggestion
  secret_id = aws_secretsmanager_secret_version.db_password.secret_id
```

I was running into a condition where, when this referenced the `aws_secretsmanager_secret`, the RDS instance would be one version behind on the secret version. I haven't been involved in the conversations around these improvements so feel free to push back on this suggestion if there's something I'm misunderstanding though!

}

resource "aws_db_instance" "example" {
  instance_class      = "db.t3.micro"
  allocated_storage   = "5"
  engine              = "postgres"
  username            = "example"
  skip_final_snapshot = true

> 

Reviewer: Extra line snuck in
```suggestion
```

  password_wo         = ephemeral.aws_secretsmanager_secret_version.db_password.secret_string

>   password_wo_version = 1

Reviewer 2: ```suggestion
  password_wo_version = aws_secretsmanager_secret_version.db_password.secret_string_wo_version
```

Reading through [your gist](https://gist.github.com/bschaatsbergen/cec07997d0d187d8e8c7889af96a8c45), should we include this as well?

}
```

> ### Deferring ephemeral resources

Reviewer: Since this is explaining the code block above, I think we should remove the header to keep the context of that code block in mind:

```suggestion
```

> If an input argument of an ephemeral resource references a value that is not yet known but will be during or after the plan, Terraform defers the resource’s execution to the apply stage instead of running it during the plan. This behavior allows Terraform to evaluate the ephemeral resource at the correct time and ensures that the resource is not executed prematurely.

Reviewer: I really like the way you gave a play-byplay of how Terraform approaches this, how would you feel about a numbered list here? I tried make this a little more pointed to the example so folks could follow along, but please feel free to disregard anything you don't like!
```suggestion
In the above example, the ephemeral resource `aws_secretsmanager_secret_version` references a `. secret_string_wo` argument. If an ephemeral resource argument references a value that Terraform cannot know yet, then Terraform defers executing that resource until the apply stage, when Terraform has the information to evaluate that argument's value. 

Terraform executes this configuration in the following order:
1. Terraform generates the ephemeral `random_password` resource
1. Terraform retrieves that `random_password` value and assigns it to `aws_secretsmanager_secret_version.secret_string_wo`
1. Terraform writes the password to the write-only `password_wo` argument of the `aws_db_instance` resource

By executing the configuration in this order, Terraform evaluates each ephemeral resource argument successfully.
```

> In the above example, the ephemeral resource `aws_secretsmanager_secret_version` references an input argument that is initially unknown. As a result, Terraform defers its execution to the apply stage, ensuring that the resource is evaluated at the correct time. 

Reviewer: (If you like my suggestion we'd remove this)
```suggestion
```

> This allows Terraform to first create the secret using the ephemeral `random_password`, then retrieve it using the ephemeral `aws_secretsmanager_secret_version` resource, and finally write the password to the write-only `password_wo` argument of the `aws_db_instance` resource.

Reviewer 3: Very minor, but creating a secret could be ambiguous (creating a value in TF versus persisting it externally). This change could make it mega explicit:

```suggestion
This allows Terraform to first create the secret in AWS Secrets Manager using the ephemeral `random_password`, then retrieve it using the ephemeral `aws_secretsmanager_secret_version` resource, and finally write the password to the write-only `password_wo` argument of the `aws_db_instance` resource.
```
